src/pstore/pstore.c

```diff
@@ -171,153 +171,148 @@ static int move_file(PStoreEntry *pe, const char *subdir) {
-                        if (!pe_id) {
+                if (plen < 6)
+                        continue;
+
+                if (startswith(pe->dirent.d_name, "dmesg-efi-")) {
+                        /* For the EFI backend, the 3 least significant digits
+                         * of record id encodes a "count" number, the next 2
+                         * least significant digits for the dmesg part (chunk)
+                         * number, and the remaining digits as the timestamp.
+                         * See linux/drivers/firmware/efi/efi-pstore.c in
+                         * efi_pstore_write(). */
+
+                        /* Extract base record id */
+                        subdir1 = memdup_suffix0(p, plen-5);
+                        if (!subdir1) {
+                                log_oom();
+                                return;
+                        }
+                        /* Extract "count" field */
+                        subdir2 = memdup_suffix0(&p[plen-3], 3);
+                        if (!subdir2) {
                                 log_oom();
                                 return;
                         }
-                } else
-                        pe_id = mfree(pe_id);
 
-                /* Now move file from pstore to archive storage */
-                move_file(pe, pe_id);
+                        /* Now move file from pstore to archive storage */
+                        move_file(pe, subdir1, subdir2);
 
-                if (dmesg_bad)
-                        continue;
+                        /* Append to the dmesg */
+                        (void)append_dmesg(pe, subdir1, subdir2);
 
-                /* If the current record id is NOT the same as the
-                 * previous record id, then start a new dmesg.txt file */
-                if (!streq_ptr(pe_id, dmesg_id)) {
-                        /* Encountered a new dmesg group, close out old one, open new one */
-                        (void) write_dmesg(dmesg, dmesg_size, dmesg_id);
-                        dmesg_size = 0;
+                        /* Cleanup */
+                        subdir1 = mfree(subdir1);
+                        subdir2 = mfree(subdir2);
+                } else
 
-                        /* now point dmesg_id to storage of pe_id */
-                        free_and_replace(dmesg_id, pe_id);
-                }
+                if (startswith(pe->dirent.d_name, "dmesg-erst-")) {
+                        /* For the ERST backend, the record is a monotonically
+                         * increasing number, seeded as a timestamp. See linux/
+                         * drivers/acpi/apei/erst.c in erst_writer().
+                         */
+                        uint64_t record_id;
+                        int r;
+
+                        r = safe_atou64(p, &record_id);
+                        if (r < 0)
+                                continue;
+                        if ((last_record_id - 1) != record_id) {
```

Comment:
The parens around the substraction are unneeded.

Reply:
done!